### PR TITLE
tty/zynq7000-uart: add uart's driver

### PR DESCRIPTION
**Pull request by HBuczynski**
## Description

Add driver implementation for uart on Zynq-7000:
- console number and baudrate is set via args,
- each uart is initialized.

JIRA: PD-191
## Motivation and Context

Needed for run user space apps on new target armv7a9-zynq7000

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (zynq-7000).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: tty/zynq7000-uart/zynq7000-uart.c [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+				baud = libtty_int_to_baudrate(atoi(optarg));
+				break;
+			case 'c':
+				console = atoi(optarg);
+				break;
+			default:
+				break;
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply tty/zynq7000-uart/zynq7000-uart.c:478]: Done

File: tty/zynq7000-uart/zynq7000-uart.c [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+int main(int argc, char **argv)
+{
+	uart_t *uart;
+	speed_t baud;
+	uint32_t port;
+	int n, c, console;
+
+	/* Default console parameters */
+	console = 1;
+	baud = B115200;
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply tty/zynq7000-uart/zynq7000-uart.c]: Done

File: tty/zynq7000-uart/zynq7000-uart.c [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+
+	return platformctl(&ctl);
+}
+
+
+static void uart_destroy(unsigned int n)
+{
+	uart_t *uart = &uart_common.uarts[n];
+
+	mmap(NULL, _PAGE_SIZE, PROT_WRITE | PROT_READ, MAP_DEVICE, OID_PHYSMEM, info[n].base);
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply tty/zynq7000-uart/zynq7000-uart.c]: Yeah, stupid mistake, definitely `munmap`. Fixed.

File: tty/zynq7000-uart/zynq7000-uart.c [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+	ctl.devclock.clkact0 = 0x1;
+	ctl.devclock.clkact1 = 0x1;
+	ctl.devclock.srcsel = 0;
+	ctl.devclock.divisor0 = 0x14;
+
+	return platformctl(&ctl);
+}
+
+
+static void uart_destroy(unsigned int n)
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply tty/zynq7000-uart/zynq7000-uart.c:379]: I haven't removed this function but changed the usage. `uart_destroy` is only used whether mutex or conditional creation is not success. It is no longer caused before libtty initialization.

File: tty/zynq7000-uart/zynq7000-uart.c [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+{
+	uart_t *uart = &uart_common.uarts[n];
+
+	munmap((void *)uart->base, _PAGE_SIZE);
+	libtty_close(&uart->tty);
+	libtty_destroy(&uart->tty);
+	resourceDestroy(uart->cond);
+	resourceDestroy(uart->lock);
```
Comment (agkaminski) [inline tty/zynq7000-uart/zynq7000-uart.c:387]: You're removing resource number 0, if we haven't created it already. So, you might init uart0 (uart->cond == 0) and then fail to init uart1 - and remove cond of uart0. That's why I don't like this function ;)

File: tty/zynq7000-uart/zynq7000-uart.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (agkaminski) [reply tty/zynq7000-uart/zynq7000-uart.c:379]: See the comment about resource destroy

File: tty/zynq7000-uart/zynq7000-uart.c [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+	uint32_t port;
+	int n, c, console = 1;
+
+	/* Get console ID and baudrate */
+	while ((c = getopt(argc, argv, "b:c:")) != -1) {
+		switch (c) {
+			case 'b':
+				if ((baud = libtty_int_to_baudrate(atoi(optarg))) == (speed_t)-1) {
+					debug("uart: wrong baudrate value\n");
+					return -1;
```
Comment (agkaminski) [inline tty/zynq7000-uart/zynq7000-uart.c:464]: As it turns out, we shouldn't return negative values from main
```suggestion
					return 1;
```

Comment (HBuczynski) [reply tty/zynq7000-uart/zynq7000-uart.c:464]: Why? I have seen in our code that it is a common practice, some instances:
https://github.com/phoenix-rtos/phoenix-rtos-devices/blob/fde0606cd220251d0c2572b34775d98796feea83/tty/uart16550/uart16550.c#L327-L330

https://github.com/phoenix-rtos/phoenix-rtos-devices/blob/fde0606cd220251d0c2572b34775d98796feea83/multi/imxrt-multi/imxrt-multi.c#L430-L433

Comment (agkaminski) [reply tty/zynq7000-uart/zynq7000-uart.c:464]: Yeah, all of these are errors to be fixed.

https://pubs.opengroup.org/onlinepubs/9699919799/functions/_exit.html
"The value of status may be 0, EXIT_SUCCESS, EXIT_FAILURE, [CX] [Option Start]  or any other value, though only the least significant 8 bits (that is, status & 0377) shall be available from wait() and waitpid(); (..)"
Only values from 0 to 255 are valid, rest of the bits are for OS purposes and are not preserved (so there is actually no way to successfully return negative value.

File: tty/zynq7000-uart/zynq7000-uart.c (hunk 5 above)
Comment (HBuczynski) [reply tty/zynq7000-uart/zynq7000-uart.c]: Yeah, you have convinced me. Removed `uart_destroy`.

**Review by agkaminski — APPROVED** [review]

**Pull request merged**